File: spaceshipgodclass.php
<?php

class SpaceshipGodClass
{
    // --- Core Ship Attributes ---
    public string $name;
    public string $shipClass;
    public float $currentHullPoints;
    public float $maxHullPoints;
    public float $currentShieldPoints;
    public float $maxShieldPoints;
    public bool $isShieldsActive = false;

    // --- Crew & Cargo ---
    public array $crewMembers = [];
    public int $maxCrewCapacity;
    public array $passengerList = [];
    public array $cargoHold = []; // item_name => quantity
    public float $currentCargoWeight = 0;
    public float $cargoCapacity;

    // --- Propulsion & Navigation ---
    public float $fuelLevel;
    public float $fuelCapacity;
    public string $engineType;
    public bool $isEngineActive = false;
    public float $currentSpeed = 0;
    public float $maxWarpFactor = 9.9;
    public float $currentWarpFactor = 0;
    public array $currentLocation = ['x' => 0, 'y' => 0, 'z' => 0];
    public ?array $destination = null;
    public string $navigationComputerStatus = "Online";
    public string $starChartsVersion = "v2.5. galactic-patch";
    public bool $autopilotEngaged = false;

    // --- Combat ---
    public array $weaponSystems;
    public array $weaponPowerLevels = [];
    public ?string $targetLock = null;
    public bool $isInCombat = false;
    public array $ammunitionCount = ["Phasers" => 1000, "Photon Torpedoes" => 50];

    // --- Sensors & Communication ---
    public string $sensorSuite;
    public ?string $longRangeScanResults = null;
    public ?string $shortRangeScanResults = null;
    public string $commsChannel = "Open Hailing Frequencies";
    public bool $isCommsEncrypted = false;
    public array $incomingMessages = [];

    // --- Life Support & Environment ---
    public string $lifeSupportLevel;
    public float $oxygenLevel = 98;
    public float $internalTemperature = 22;
    public string $gravityPlatingStatus = "Active";
    public float $wasteRecyclingEfficiency = 85;

    // --- Miscellaneous ---
    public string $coffeeMachineModel;
    public float $coffeeLevel = 100;
    public bool $isHolodeckInUse = false;
    public ?string $holodeckProgramRunning = null;
    public array $shipsLog = [];
    public bool $selfDestructSequenceInitiated = false;
    public string $selfDestructCode = "000-DESTRUCT-0";
    public string $insurancePolicyNumber = "GCS-1701-OMG";

    public function __construct(
        string $name, string $shipClass, float $maxHullPoints, float $maxShieldPoints,
        int $maxCrewCapacity, float $cargoCapacity, float $fuelCapacity,
        string $engineType, array $weaponSystems, string $sensorSuite,
        string $lifeSupportLevel, string $coffeeMachineModel
    ) {
        echo "Constructing the mighty {$name}, a {$shipClass}-class vessel!\n";

        $this->name = $name;
        $this->shipClass = $shipClass;
        $this->currentHullPoints = $maxHullPoints;
        $this->maxHullPoints = $maxHullPoints;
        $this->currentShieldPoints = $maxShieldPoints;
        $this->maxShieldPoints = $maxShieldPoints;
        $this->maxCrewCapacity = $maxCrewCapacity;
        $this->cargoCapacity = $cargoCapacity;
        $this->fuelLevel = $fuelCapacity;
        $this->fuelCapacity = $fuelCapacity;
        $this->engineType = $engineType;
        $this->weaponSystems = $weaponSystems;
        foreach ($weaponSystems as $weapon) {
            $this->weaponPowerLevels[$weapon] = 100;
        }
        $this->sensorSuite = $sensorSuite;
        $this->lifeSupportLevel = $lifeSupportLevel;
        $this->coffeeMachineModel = $coffeeMachineModel;

        echo "{$name} construction complete. All systems... probably online.\n";
        $this->logEntry("Ship constructed.");
    }

    // --- Logging ---
    public function logEntry(string $message): void
    {
        $timestamp = date("c"); // ISO 8601 format
        $logMessage = "[{$timestamp}] - {$message}";
        $this->shipsLog[] = $logMessage;
        echo "Log: {$message}\n";
    }

    // --- Hull & Shields ---
    public function takeDamage(float $amount): void
    {
        if (!$this->isShieldsActive || $this->currentShieldPoints <= 0) {
            $this->currentHullPoints -= $amount;
            $this->logEntry("Took {$amount} hull damage.");
            if ($this->currentHullPoints <= 0) {
                $this->logEntry("HULL BREACH! SHIP DESTROYED!");
                $this->currentHullPoints = 0;
                echo "💥💥💥 BOOM! 💥💥💥\n";
            }
        } else {
            $shieldDamage = min($this->currentShieldPoints, $amount);
            $this->currentShieldPoints -= $shieldDamage;
            $remainingDamage = $amount - $shieldDamage;
            $this->logEntry("Shields absorbed {$shieldDamage} damage.");
            if ($remainingDamage > 0) {
                $this->currentHullPoints -= $remainingDamage;
                $this->logEntry("Took {$remainingDamage} hull damage through shields.");
            }
            if ($this->currentShieldPoints <= 0) {
                $this->logEntry("Shields are down!");
                $this->isShieldsActive = false;
            }
        }
    }

    public function repairHull(float $amount): void
    {
        $this->currentHullPoints = min($this->maxHullPoints, $this->currentHullPoints + $amount);
        $this->logEntry("Hull repaired by {$amount} points. Current: {$this->currentHullPoints}");
    }

    public function toggleShields(bool $state): void
    {
        if ($state && $this->currentShieldPoints > 0) {
            $this->isShieldsActive = true;
            $this->logEntry("Shields up!");
        } else {
            $this->isShieldsActive = false;
            $this->logEntry("Shields down!");
        }
    }

    public function rechargeShields(float $amount): void
    {
        $this->currentShieldPoints = min($this->maxShieldPoints, $this->currentShieldPoints + $amount);
        $this->logEntry("Shields recharged to {$this->currentShieldPoints}.");
    }

    // --- Crew & Cargo Management ---
     public function addCrewMember(string $name, string $position): bool
     {
        if (count($this->crewMembers) < $this->maxCrewCapacity) {
            $this->crewMembers[] = ["name" => $name, "position" => $position];
            $this->logEntry("Crew member {$name} ({$position}) has come aboard.");
            return true;
        } else {
            $this->logEntry("Cannot add crew, capacity reached.");
            return false;
        }
     }

     public function loadCargo(string $itemName, int $quantity, float $weightPerUnit): bool
     {
        $totalWeight = $quantity * $weightPerUnit;
        if ($this->currentCargoWeight + $totalWeight <= $this->cargoCapacity) {
            $this->cargoHold[$itemName] = ($this->cargoHold[$itemName] ?? 0) + $quantity;
            $this->currentCargoWeight += $totalWeight;
            $this->logEntry("Loaded {$quantity} units of {$itemName}.");
            return true;
        } else {
            $this->logEntry("Cannot load {$itemName}, insufficient cargo space.");
            return false;
        }
     }

    // --- Engine & Navigation ---
    public function startEngine(): void
    {
        if ($this->fuelLevel > 0) {
            $this->isEngineActive = true;
            $this->logEntry("Engines started.");
        } else {
            $this->logEntry("Cannot start engines, no fuel.");
        }
    }

    public function stopEngine(): void
    {
        $this->isEngineActive = false;
        $this->currentSpeed = 0;
        $this->currentWarpFactor = 0;
        $this->logEntry("Engines stopped.");
    }

     public function engageWarp(float $factor): void
     {
        if ($this->isEngineActive && $factor <= $this->maxWarpFactor) {
            $this->currentWarpFactor = $factor;
            $this->currentSpeed = 0;
            $this->logEntry("Warp factor {$factor} engaged!");
        } elseif (!$this->isEngineActive) {
            $this->logEntry("Cannot engage warp, engines offline.");
        } else {
            $this->logEntry("Cannot engage warp factor {$factor}, exceeds maximum.");
        }
     }

     public function setCourse(float $x, float $y, float $z): void
     {
        $this->destination = ['x' => $x, 'y' => $y, 'z' => $z];
        $this->logEntry("Course set for coordinates: ({$x}, {$y}, {$z}).");
     }

    // ... (Other methods would be translated similarly) ...

    // --- Coffee & Holodeck ---
    public function brewCoffee(string $strength = "Strong"): ?string
    {
        if ($this->coffeeLevel > 5) {
            $this->coffeeLevel -= 5;
            $this->logEntry("Brewing a {$strength} coffee. Ahh, essential for deep space travel.");
            return "☕";
        } else {
            $this->logEntry("CRITICAL ERROR: Coffee depleted! Morale dropping!");
            return null;
        }
    }

    // --- Status Report ---
    public function getFullStatusReport(): void
    {
        echo "\n" . str_repeat("=", 50) . "\n";
        echo "       STATUS REPORT: {$this->name}       \n";
        echo str_repeat("=", 50) . "\n";
        echo "Class: {$this->shipClass}\n";
        echo "Hull: {$this->currentHullPoints}/{$this->maxHullPoints}\n";
        $shieldsActive = $this->isShieldsActive ? 'Active' : 'Inactive';
        echo "Shields: {$this->currentShieldPoints}/{$this->maxShieldPoints} (Active: {$shieldsActive})\n";
        echo "Crew: " . count($this->crewMembers) . "/{$this->maxCrewCapacity}\n";
        echo "Cargo: {$this->currentCargoWeight}/{$this->cargoCapacity} tons\n";
        echo "Fuel: " . number_format($this->fuelLevel, 2) . "/{$this->fuelCapacity}\n";
        $engineActive = $this->isEngineActive ? 'Active' : 'Inactive';
        echo "Engines: {$this->engineType} (Active: {$engineActive})\n";
        echo "Speed: Warp {$this->currentWarpFactor} / Impulse {$this->currentSpeed}\n";
        echo "Location: ({$this->currentLocation['x']}, {$this->currentLocation['y']}, {$this->currentLocation['z']})\n";
        $dest = $this->destination ? "({$this->destination['x']}, {$this->destination['y']}, {$this->destination['z']})" : "None";
        echo "Destination: {$dest}\n";
        echo "Weapons: " . implode(', ', $this->weaponSystems) . "\n";
        $target = $this->targetLock ?? "None";
        echo "Target: {$target}\n";
        echo "Life Support: {$this->lifeSupportLevel} (O2: {$this->oxygenLevel}%, Temp: {$this->internalTemperature}°C)\n";
        echo "Coffee Level: {$this->coffeeLevel}%\n";
        $holodeck = $this->isHolodeckInUse ? 'In use' : 'Idle';
        echo "Holodeck: {$holodeck}\n";
        $selfDestruct = $this->selfDestructSequenceInitiated ? 'ARMED' : 'Safe';
        echo "Self Destruct: {$selfDestruct}\n";
        echo str_repeat("=", 50) . "\n\n";
    }
}

// --- Example Usage ---
$myShip = new SpaceshipGodClass(
    "The Colossus PHP",
    "Server-Side-Dreadnought",
    5000, 10000, 1015, 50000, 100000,
    "PHP-Powered Warp Drive",
    ["Phasers", "Photon Torpedoes", "SQL Injectors"], // Beware!
    "Zend Engine Sensor Array",
    "Session-Based",
    "Nespresso Pro"
);

$myShip->addCrewMember("Rasmus Lerdorf", "Captain");
$myShip->addCrewMember("Zeev Suraski", "First Officer");
$myShip->loadCargo("Frameworks", 500, 5);
$myShip->toggleShields(true);
$myShip->startEngine();
$myShip->setCourse(100, 250, -50);
$myShip->engageWarp(7);
$myShip->takeDamage(3500);
$myShip->brewCoffee("Espresso");
$myShip->getFullStatusReport();

File: spaceshipgodclass.py
class SpaceshipGodClass:
    """
    Behold! The epitome of a God Class - our Spaceship!
    It knows everything, it does everything. It navigates, fights,
    manages life support, brews coffee, and probably even questions
    its own existence in its spare CPU cycles.
    """

    def __init__(self, name, ship_class, max_hull_points, max_shield_points,
                 max_crew_capacity, cargo_capacity, fuel_capacity,
                 engine_type, weapon_systems, sensor_suite,
                 life_support_level, coffee_machine_model):
        """
        Our constructor - because even a God Class needs a beginning.
        We initialize *everything* here.
        """
        print(f"Constructing the mighty {name}, a {ship_class}-class vessel!")

        # --- Core Ship Attributes ---
        self.name = name
        self.ship_class = ship_class
        self.current_hull_points = max_hull_points
        self.max_hull_points = max_hull_points
        self.current_shield_points = max_shield_points
        self.max_shield_points = max_shield_points
        self.is_shields_active = False

        # --- Crew & Cargo ---
        self.crew_members = []
        self.max_crew_capacity = max_crew_capacity
        self.passenger_list = []
        self.cargo_hold = {} # item_name: quantity
        self.current_cargo_weight = 0
        self.cargo_capacity = cargo_capacity # Assuming in tons

        # --- Propulsion & Navigation ---
        self.fuel_level = fuel_capacity
        self.fuel_capacity = fuel_capacity
        self.engine_type = engine_type
        self.is_engine_active = False
        self.current_speed = 0
        self.max_warp_factor = 9.9 # Why not?
        self.current_warp_factor = 0
        self.current_location = (0, 0, 0) # x, y, z coordinates
        self.destination = None
        self.navigation_computer_status = "Online"
        self.star_charts_version = "v2.5. galactic-patch"
        self.autopilot_engaged = False

        # --- Combat ---
        self.weapon_systems = weapon_systems # List of weapon objects? No, just names for now!
        self.weapon_power_levels = {weapon: 100 for weapon in weapon_systems}
        self.target_lock = None
        self.is_in_combat = False
        self.ammunition_count = {"Phasers": 1000, "Photon Torpedoes": 50}

        # --- Sensors & Communication ---
        self.sensor_suite = sensor_suite
        self.long_range_scan_results = None
        self.short_range_scan_results = None
        self.comms_channel = "Open Hailing Frequencies"
        self.is_comms_encrypted = False
        self.incoming_messages = []

        # --- Life Support & Environment ---
        self.life_support_level = life_support_level # e.g., "Nominal"
        self.oxygen_level = 98 # Percent
        self.internal_temperature = 22 # Celsius
        self.gravity_plating_status = "Active"
        self.waste_recycling_efficiency = 85 # Percent

        # --- Miscellaneous (Because why not?) ---
        self.coffee_machine_model = coffee_machine_model
        self.coffee_level = 100 # Percent
        self.is_holodeck_in_use = False
        self.holodeck_program_running = None
        self.ships_log = []
        self.self_destruct_sequence_initiated = False
        self.self_destruct_code = "000-DESTRUCT-0"
        self.insurance_policy_number = "GCS-1701-OMG"

        print(f"{name} construction complete. All systems... probably online.")
        self.log_entry("Ship constructed.")

    # --- Logging ---
    def log_entry(self, message):
        """Adds an entry to the ship's log."""
        import datetime
        timestamp = datetime.datetime.now().isoformat()
        self.ships_log.append(f"[{timestamp}] - {message}")
        print(f"Log: {message}")

    # --- Hull & Shields ---
    def take_damage(self, amount):
        """Handles incoming damage."""
        if not self.is_shields_active or self.current_shield_points <= 0:
            self.current_hull_points -= amount
            self.log_entry(f"Took {amount} hull damage.")
            if self.current_hull_points <= 0:
                self.log_entry("HULL BREACH! SHIP DESTROYED!")
                self.current_hull_points = 0
                print("💥💥💥 BOOM! 💥💥💥")
        else:
            shield_damage = min(self.current_shield_points, amount)
            self.current_shield_points -= shield_damage
            remaining_damage = amount - shield_damage
            self.log_entry(f"Shields absorbed {shield_damage} damage.")
            if remaining_damage > 0:
                self.current_hull_points -= remaining_damage
                self.log_entry(f"Took {remaining_damage} hull damage through shields.")
            if self.current_shield_points <= 0:
                self.log_entry("Shields are down!")
                self.is_shields_active = False

    def repair_hull(self, amount):
        """Repairs the hull."""
        self.current_hull_points = min(self.max_hull_points, self.current_hull_points + amount)
        self.log_entry(f"Hull repaired by {amount} points. Current: {self.current_hull_points}")

    def toggle_shields(self, state):
        """Activates or deactivates shields."""
        if state and self.current_shield_points > 0:
            self.is_shields_active = True
            self.log_entry("Shields up!")
        else:
            self.is_shields_active = False
            self.log_entry("Shields down!")

    def recharge_shields(self, amount):
        """Recharges the shields."""
        self.current_shield_points = min(self.max_shield_points, self.current_shield_points + amount)
        self.log_entry(f"Shields recharged to {self.current_shield_points}.")

    # --- Crew & Cargo Management ---
    def add_crew_member(self, name, position):
        """Adds a crew member."""
        if len(self.crew_members) < self.max_crew_capacity:
            self.crew_members.append({"name": name, "position": position})
            self.log_entry(f"Crew member {name} ({position}) has come aboard.")
            return True
        else:
            self.log_entry("Cannot add crew, capacity reached.")
            return False

    def load_cargo(self, item_name, quantity, weight_per_unit):
        """Loads cargo into the hold."""
        total_weight = quantity * weight_per_unit
        if self.current_cargo_weight + total_weight <= self.cargo_capacity:
            self.cargo_hold[item_name] = self.cargo_hold.get(item_name, 0) + quantity
            self.current_cargo_weight += total_weight
            self.log_entry(f"Loaded {quantity} units of {item_name}.")
            return True
        else:
            self.log_entry(f"Cannot load {item_name}, insufficient cargo space.")
            return False

    def jettison_cargo(self, item_name, quantity, weight_per_unit):
        """Jettisons cargo."""
        if item_name in self.cargo_hold and self.cargo_hold[item_name] >= quantity:
            self.cargo_hold[item_name] -= quantity
            self.current_cargo_weight -= quantity * weight_per_unit
            self.log_entry(f"Jettisoned {quantity} units of {item_name}.")
            if self.cargo_hold[item_name] == 0:
                del self.cargo_hold[item_name]
            return True
        else:
            self.log_entry(f"Cannot jettison {item_name}, not enough in hold.")
            return False

    # --- Engine & Navigation ---
    def start_engine(self):
        """Starts the main engines."""
        if self.fuel_level > 0:
            self.is_engine_active = True
            self.log_entry("Engines started.")
        else:
            self.log_entry("Cannot start engines, no fuel.")

    def stop_engine(self):
        """Stops the main engines."""
        self.is_engine_active = False
        self.current_speed = 0
        self.current_warp_factor = 0
        self.log_entry("Engines stopped.")

    def set_impulse_speed(self, speed_fraction):
        """Sets the impulse speed (sub-light)."""
        if self.is_engine_active:
            self.current_speed = speed_fraction # Assuming fraction of light speed
            self.current_warp_factor = 0
            self.log_entry(f"Impulse speed set to {speed_fraction}.")
        else:
            self.log_entry("Cannot set speed, engines offline.")

    def engage_warp(self, factor):
        """Engages the warp drive."""
        if self.is_engine_active and factor <= self.max_warp_factor:
            self.current_warp_factor = factor
            self.current_speed = 0 # Or some representation of FTL
            self.log_entry(f"Warp factor {factor} engaged!")
        elif not self.is_engine_active:
            self.log_entry("Cannot engage warp, engines offline.")
        else:
            self.log_entry(f"Cannot engage warp factor {factor}, exceeds maximum.")

    def set_course(self, x, y, z):
        """Sets a new destination."""
        self.destination = (x, y, z)
        self.log_entry(f"Course set for coordinates: {self.destination}.")

    def travel(self, time_delta):
        """Simulates travel over a period of time."""
        if not self.is_engine_active or self.destination is None:
            self.log_entry("Cannot travel, no active engines or destination.")
            return

        # Extremely simplified travel calculation - THIS IS NOT REAL ASTROPHYSICS!
        if self.current_warp_factor > 0:
            distance_travelled = self.current_warp_factor**3 * time_delta # Arbitrary formula
        else:
            distance_travelled = self.current_speed * time_delta

        # Update fuel - also very arbitrary
        fuel_consumed = (self.current_warp_factor + self.current_speed) * time_delta * 5
        self.fuel_level = max(0, self.fuel_level - fuel_consumed)

        # Move towards destination (super simple linear movement)
        dx = self.destination[0] - self.current_location[0]
        dy = self.destination[1] - self.current_location[1]
        dz = self.destination[2] - self.current_location[2]
        distance_to_go = (dx**2 + dy**2 + dz**2)**0.5

        if distance_to_go <= distance_travelled:
            self.current_location = self.destination
            self.destination = None
            self.stop_engine()
            self.log_entry(f"Arrived at destination: {self.current_location}. Engines stopped.")
        else:
            move_fraction = distance_travelled / distance_to_go
            new_x = self.current_location[0] + dx * move_fraction
            new_y = self.current_location[1] + dy * move_fraction
            new_z = self.current_location[2] + dz * move_fraction
            self.current_location = (new_x, new_y, new_z)
            self.log_entry(f"Travelled {distance_travelled} units. New location: {self.current_location}. Fuel: {self.fuel_level:.2f}")

        if self.fuel_level <= 0:
            self.log_entry("WARNING: Fuel depleted! Adrift in space.")
            self.stop_engine()

    # --- Combat Functions ---
    def fire_weapon(self, weapon_name, target):
        """Fires a specified weapon at a target."""
        if weapon_name in self.weapon_systems and self.weapon_power_levels[weapon_name] > 0:
            if self.ammunition_count.get(weapon_name, 1) > 0:
                self.log_entry(f"Firing {weapon_name} at {target}!")
                # In a real (bad) god class, we'd probably try to resolve
                # the hit and damage right here...
                self.ammunition_count[weapon_name] = self.ammunition_count.get(weapon_name, 1) -1
                print(f"Pew pew! {weapon_name} fired!")
                return True
            else:
                self.log_entry(f"Cannot fire {weapon_name}, no ammunition.")
                return False
        else:
            self.log_entry(f"Cannot fire {weapon_name}, system offline or not equipped.")
            return False

    def lock_target(self, target_signature):
        """Locks onto an enemy target."""
        self.target_lock = target_signature
        self.log_entry(f"Target {target_signature} locked.")

    # --- Sensor & Comms ---
    def perform_scan(self, scan_type="Long"):
        """Performs a sensor scan."""
        if scan_type == "Long":
            self.long_range_scan_results = f"Detected nebulas and a distant G-type star at {datetime.datetime.now()}"
            self.log_entry("Long-range scan completed.")
            return self.long_range_scan_results
        else:
            self.short_range_scan_results = f"Detected asteroids and a small Class-M planet nearby at {datetime.datetime.now()}"
            self.log_entry("Short-range scan completed.")
            return self.short_range_scan_results

    def send_hail(self, message, target_ship):
        """Sends a communication message."""
        self.log_entry(f"Hailing {target_ship}: '{message}'")
        # Imagine complex comms logic here...
        print(f"Comms: '{message}' sent.")

    # --- Life Support ---
    def adjust_oxygen(self, level):
        """Adjusts oxygen levels."""
        self.oxygen_level = level
        self.log_entry(f"Oxygen adjusted to {level}%.")

    def adjust_temperature(self, temp):
        """Adjusts internal temperature."""
        self.internal_temperature = temp
        self.log_entry(f"Temperature set to {temp}°C.")

    # --- Coffee & Holodeck (Critical Systems!) ---
    def brew_coffee(self, strength="Strong"):
        """Brews a cup of coffee."""
        if self.coffee_level > 5:
            self.coffee_level -= 5
            self.log_entry(f"Brewing a {strength} coffee. Ahh, essential for deep space travel.")
            return "☕"
        else:
            self.log_entry("CRITICAL ERROR: Coffee depleted! Morale dropping!")
            return None

    def start_holodeck_program(self, program_name):
        """Starts a holodeck program."""
        if not self.is_holodeck_in_use:
            self.is_holodeck_in_use = True
            self.holodeck_program_running = program_name
            self.log_entry(f"Holodeck program '{program_name}' initiated. Enjoy!")
        else:
            self.log_entry("Holodeck already in use.")

    def stop_holodeck_program(self):
        """Stops the current holodeck program."""
        if self.is_holodeck_in_use:
            self.log_entry(f"Holodeck program '{self.holodeck_program_running}' ended.")
            self.is_holodeck_in_use = False
            self.holodeck_program_running = None
        else:
            self.log_entry("Holodeck is not currently active.")

    # --- Self Destruct ---
    def initiate_self_destruct(self, code):
        """Initiates the self-destruct sequence."""
        if code == self.self_destruct_code:
            self.self_destruct_sequence_initiated = True
            self.log_entry("WARNING: Self-destruct sequence initiated! T-minus 5 minutes.")
            print("🚨🚨🚨 SELF DESTRUCT INITIATED 🚨🚨🚨")
        else:
            self.log_entry("Invalid self-destruct code. Sequence aborted.")

    def abort_self_destruct(self):
        """Aborts the self-destruct sequence."""
        if self.self_destruct_sequence_initiated:
            self.self_destruct_sequence_initiated = False
            self.log_entry("Self-destruct sequence aborted. Phew!")
        else:
            self.log_entry("No self-destruct sequence active.")

    # --- Status Report (Because of course) ---
    def get_full_status_report(self):
        """Prints a massive status report."""
        print("\n" + "="*50)
        print(f"       STATUS REPORT: {self.name}       ")
        print("="*50)
        print(f"Class: {self.ship_class}")
        print(f"Hull: {self.current_hull_points}/{self.max_hull_points}")
        print(f"Shields: {self.current_shield_points}/{self.max_shield_points} (Active: {self.is_shields_active})")
        print(f"Crew: {len(self.crew_members)}/{self.max_crew_capacity}")
        print(f"Cargo: {self.current_cargo_weight}/{self.cargo_capacity} tons")
        print(f"Fuel: {self.fuel_level:.2f}/{self.fuel_capacity}")
        print(f"Engines: {self.engine_type} (Active: {self.is_engine_active})")
        print(f"Speed: Warp {self.current_warp_factor} / Impulse {self.current_speed}")
        print(f"Location: {self.current_location}")
        print(f"Destination: {self.destination}")
        print(f"Weapons: {', '.join(self.weapon_systems)}")
        print(f"Target: {self.target_lock}")
        print(f"Life Support: {self.life_support_level} (O2: {self.oxygen_level}%, Temp: {self.internal_temperature}°C)")
        print(f"Coffee Level: {self.coffee_level}%")
        print(f"Holodeck: {'In use' if self.is_holodeck_in_use else 'Idle'}")
        print(f"Self Destruct: {'ARMED' if self.self_destruct_sequence_initiated else 'Safe'}")
        print("="*50 + "\n")

# --- Example Usage ---
if __name__ == "__main__":
    # Create our behemoth
    my_ship = SpaceshipGodClass(
        name="The Colossus",
        ship_class="Galaxy-Dreadnought-Espresso-Carrier",
        max_hull_points=5000,
        max_shield_points=10000,
        max_crew_capacity=1015,
        cargo_capacity=50000,
        fuel_capacity=100000,
        engine_type="Dilithium-Powered Warp/Impulse Fusion",
        weapon_systems=["Phasers", "Photon Torpedoes", "Disruptor Cannons"],
        sensor_suite="Argus Array Mk V",
        life_support_level="Optimized",
        coffee_machine_model="BaristaBot 9000"
    )

    # Let's make it do *everything*!
    my_ship.add_crew_member("Jean-Luc Picard", "Captain")
    my_ship.add_crew_member("William Riker", "First Officer")
    my_ship.add_crew_member("Data", "Operations Officer")

    my_ship.load_cargo("Tribbles", 10000, 0.1) # Oh no!
    my_ship.load_cargo("Dilithium Crystals", 50, 100)

    my_ship.toggle_shields(True)
    my_ship.start_engine()
    my_ship.set_course(100, 250, -50)
    my_ship.engage_warp(7)
    my_ship.travel(10) # Travel for 10 time units

    my_ship.lock_target("Hostile Bird-of-Prey")
    my_ship.fire_weapon("Phasers", "Hostile Bird-of-Prey")
    my_ship.fire_weapon("Photon Torpedoes", "Hostile Bird-of-Prey")

    my_ship.take_damage(3500) # We took a hit!
    my_ship.recharge_shields(1000)

    my_ship.perform_scan("Short")
    my_ship.send_hail("This is Captain Picard. Surrender now!", "Hostile Bird-of-Prey")

    my_ship.brew_coffee("Earl Grey, Hot")
    my_ship.start_holodeck_program("Dixon Hill - Private Investigator")

    my_ship.get_full_status_report()

    # Maybe we shouldn't have loaded those tribbles...
    # my_ship.initiate_self_destruct("000-DESTRUCT-0")
